Add com.iterable/system_webhook/jsonschema/1-0-0 (close #1204)

matus-tomlein opened the pull request:
Issue #1024 

The schema was produced based on example payloads in the [Iterable system webhook documentation](https://support.iterable.com/hc/en-us/articles/208013936-System-Webhooks-) and events that I was able to trigger manually in a test account on Iterable. Unfortunately, there is no documentation that would list all the properties and their types so I can't be 100% sure about the schema. We also didn't get a response from Iterable after we asked them to confirm the properties and types.

I decided to leave out the `transactionalData` property that I saw conflicting examples for and couldn't find out it's type. We may add it in a future version of the schema when we are more certain about it. Other properties I am fairly certain about and would be surprised if they turned out to have a different type (although I can't rule that out).

I created a testing repository where I collected the examples of the events and created a script that sends the events to a Micro instance to validate that the schema supports them: https://github.com/snowplow-incubator/snowplow-iterable-webhook-schema

snowplowcla commented:
Thanks for your pull request. Is this your first contribution to a Snowplow open source project? Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).

:memo: Please visit https://docs.snowplowanalytics.com/docs/contributing/contributor-license-agreement/ to learn more and sign.

Once you've signed, please reply here (e.g. I signed it!) and we'll verify. Thanks.

matus-tomlein commented:
After a response from Iterable about the property types, I became even more confused. They said that `transactionalData` and `payload` are string properties which conflicts with the examples published in their docs. I decided to remove even the `payload` property from the schema here since I am not sure about its type. I still think that it'd be better to add these properties in future schema versions rather than risking schema violations.

adatzer commented on schemas/com.iterable/system_webhook/jsonschema/1-0-0:
```diff
@@ -0,0 +1,643 @@
+                  "string",
+                  "null"
+                ],
+                "description": "Reason (e.g., 'DuplicateMarketingMessage')"
```
also here - what about a maxLength?

adatzer commented on schemas/com.iterable/system_webhook/jsonschema/1-0-0:
```diff
@@ -0,0 +1,643 @@
+                  "string",
+                  "null"
+                ],
+                "description": "SMS message text"
```
another one hard to restict, but maybe there is some limit for sms length? maybe something relatively big sounds safe? what do you think?

adatzer commented on schemas/com.iterable/system_webhook/jsonschema/1-0-0:
```diff
@@ -0,0 +1,643 @@
+                      "string",
+                      "null"
+                    ],
+                    "description": "SMS provider response message from the event (e.g., 'The requested resource was not found'"
```
here too - maybe some large value for `maxLength`?

adatzer commented on schemas/com.iterable/system_webhook/jsonschema/1-0-0:
```diff
@@ -0,0 +1,643 @@
+                      "string",
+                      "null"
+                    ],
+                    "description": "SMS provider response addition info from the event (e.g, a URL)"
```
here too - maybe some large value for `maxLength`?

adatzer commented on schemas/com.iterable/system_webhook/jsonschema/1-0-0:
```diff
@@ -0,0 +1,643 @@
+                "type": [
+                  "number",
+                  "null"
+                ],
```
`["integer", "null"]` here too?

adatzer commented on schemas/com.iterable/system_webhook/jsonschema/1-0-0:
```diff
@@ -0,0 +1,643 @@
+                  "string",
+                  "null"
+                ],
+                "description": "Web push notification body"
```
here too - maybe some large value for `maxLength`?

adatzer commented on schemas/com.iterable/system_webhook/jsonschema/1-0-0:
```diff
@@ -0,0 +1,643 @@
+                  "string",
+                  "null"
+                ],
+                "description": "Web push notification icon"
```
not sure how large this string could be, but maybe there is a safe maxLength?

adatzer commented on schemas/com.iterable/system_webhook/jsonschema/1-0-0:
```diff
@@ -0,0 +1,643 @@
+                  "null"
+                ],
+                "minimum": 0,
+                "description": "Reference to the workflow which the event originated (e.g., 53505)"
```
also type `["integer", "null"]` here?

matus-tomlein commented:
Thanks a lot @adatzer for the thorough and really useful review!

I addressed all the issues:
- changed numbers to integers – I actually thought that there is no integer type
- added maxLength to all strings – usually 4096 except for the `inAppBody` HTML where I used 32700 and `email` fields where I used 320
- added "format": "email" for email fields
- I also added some large "maximum" limits for the integers

matus-tomlein commented on schemas/com.iterable/system_webhook/jsonschema/1-0-0:
```diff
@@ -0,0 +1,643 @@
+                "type": [
+                  "string",
+                  "null"
+                ],
```
Yeah, I was also thinking about that. Unfortunately, it seems that Iterable is using a different date-time format...

matus-tomlein commented:
The commit was cherry picked into release 134 branch with release PR #1210
